### PR TITLE
Floating-point operations logging in trainer

### PR DESCRIPTION
**Pull request by TevenLeScao**
First of two PRs to implement #4847 :

 - logging loss vs floating-point operations
 - using the results for scaling laws analysis

This directly logs floating-point operations in wandb and comet, and creates a `log_history.json` file with training metrics. To do so, it adds methods to `PretrainedModel` to count parameters with and without embeddings, and the number of floating-point operations. It also has a few Trainer fixes, most importantly averaging the eval loss across processes rather than logging the one in process 0, and a bug with checkpoint folder creation.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -307,9 +293,77 @@ def _convert_head_mask_to_5d(self, head_mask, num_hidden_layers):
...
 
+    def num_parameters(self, only_trainable: bool = False, no_embeddings: bool = False) -> int:
+        """
+        Get number of (optionally, trainable or non-embeddings) parameters in the module.
+
+        Args:
+            only_trainable (:obj:`bool`, `optional`, defaults to :obj:`False`):
+                Whether or not to return only the number of trainable parameters
+
+            no_embeddings (:obj:`bool`, `optional`, defaults to :obj:`False`):
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:307]: Not super fond of that name. `exclude_embeddings`?

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -307,9 +293,77 @@ def _convert_head_mask_to_5d(self, head_mask, num_hidden_layers):
...
+
+        def parameter_filter(x):
+            return (x.requires_grad or not only_trainable) and not (
+                isinstance(x, torch.nn.Embedding) and no_embeddings
+            )
+
+        params = filter(parameter_filter, self.parameters()) if only_trainable else self.parameters()
+        return sum(p.numel() for p in params)
+
+    def estimate_tokens(self, input_dict: Dict[str, Union[torch.Tensor, Any]]):
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:322]: It would be great to add the return type annotation (`Tuple[int,int]` ?)

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -307,9 +293,77 @@ def _convert_head_mask_to_5d(self, head_mask, num_hidden_layers):
...
+        """
+        Helper function to estimate the batch size and sequence length from the model inputs. Returned batch size is the
+        first dimension of input tensors, returned sequence length is the sum of the second dimensions of all input
+        tensors.
+
+        Args:
+            inputs (:obj:`dict`): The model inputs.
+
+        Returns:
+            seed (:obj:`tuple`): The batch size and sequence length.
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:332]: Why seed?
```suggestion
            :obj:`Tuple[int, int]`: The batch size and sequence length.
```

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -307,9 +293,77 @@ def _convert_head_mask_to_5d(self, head_mask, num_hidden_layers):
...
+        """
+        token_inputs = [tensor for key, tensor in input_dict.items() if "input" in key]
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:334]: I'm confused. This does not return what the docstrings say it's supposed to return.

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -307,9 +293,77 @@ def _convert_head_mask_to_5d(self, head_mask, num_hidden_layers):
...
+        if doing long-range modeling with very high sequence lengths.
+
+        Args:
+            batch_size (:obj:`int`):
+                The batch size for the forward pass.
+
+            sequence_length (:obj:`int`):
+                The number of tokens in each line of the batch.
+
+            no_embeddings (:obj:`bool`, `optional`, defaults to :obj:`True`):
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:358]: Same comment as above.

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -307,9 +293,77 @@ def _convert_head_mask_to_5d(self, head_mask, num_hidden_layers):
...
+            warnings.warn(
+                "Could not estimate the number of tokens of the input, floating-point operations will not be computed"
+            )
+            return 0
+
+    def floating_point_ops(self, input_dict: Dict[str, Union[torch.Tensor, Any]], no_embeddings: bool = True) -> int:
+        """
+        Get number of (optionally, non-embeddings) floating-point operations for the forward and backward passes of a
+        batch with this transformer model. Default approximation neglects the quadratic dependency on the number of
+        tokens (valid if 12 * d_model << sequence_length) as laid out in https://arxiv.org/pdf/2001.08361.pdf section
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:347]:
```suggestion
        tokens (valid if :obj:`12 * d_model << sequence_length`) as laid out in `this paper <https://arxiv.org/pdf/2001.08361.pdf>`__ section
```

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -690,6 +701,12 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
 
                 tr_loss += self.training_step(model, inputs)
 
+                try:
```
Comment (sgugger) [inline src/transformers/trainer.py:704]: Can we make a cleaner test with `isinstance(model, nn.DataParallel)`?

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -1092,13 +1118,24 @@ def _save_tpu(self, output_dir: Optional[str] = None):
...
+                self.log_history, open(os.path.join(output_dir, "log_history.json"), "w"), indent=2, ensure_ascii=False
+            )
 
         # Save a trained model and configuration using `save_pretrained()`.
         # They can then be reloaded using `from_pretrained()`
         if not isinstance(self.model, PreTrainedModel):
             raise ValueError("Trainer.model appears to not be a PreTrainedModel")
 
         xm.rendezvous("saving_checkpoint")
+        # Storing the number of floating-point operations that went into the model
```
Comment (sgugger) [inline src/transformers/trainer.py:1131]: Those 7 lines are duplicated, maybe put them in a private method to refactor a bit?

File: src/transformers/trainer.py [hunk 9]
```diff
@@ -1297,17 +1350,32 @@ def prediction_loop(
...
         output_tensors = [tensor.clone() for _ in range(torch.distributed.get_world_size())]
         torch.distributed.all_gather(output_tensors, tensor)
+        concat = torch.cat(output_tensors, dim=0)
+
+        # truncate the dummy elements added by SequentialDistributedSampler
+        if num_total_examples is not None:
+            concat = concat[:num_total_examples]
+        return concat
+
+    def distributed_broadcast_scalars(
```
Comment (sgugger) [inline src/transformers/trainer.py:1365]: This doesn't seem to use self (distributed_concat neither) so maybe those two methods should be functions?

File: src/transformers/trainer.py [hunk 10]
```diff
@@ -1354,3 +1422,25 @@ def prediction_step(
         if labels is not None:
             labels = labels.detach()
         return (loss, logits.detach(), labels)
+
+    def floating_point_ops(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]]):
+        """
+        For models with a `floating_point_ops` method (e.g. models that inherit from `PretrainedModel`), uses
```
Comment (sgugger) [inline src/transformers/trainer.py:1428]:
```suggestion
        For models that inherit from :class:`~transformers.PretrainedModel`, uses
```
I don't think it's worth mentioning the method since people are supposed to use our models for Trainer, that all inherit from `PretrainedModel`.

File: src/transformers/trainer.py [hunk 11]
```diff
@@ -1354,3 +1422,25 @@ def prediction_step(
...
+        that method to compute the number of floating point operations for every backward + forward pass. If using
+        another model, either implement such a method in the model or override this method.
+
+        Args:
+            model (:obj:`nn.Module`):
```
Comment (sgugger) [inline src/transformers/trainer.py:1433]: Can't we use `self.model`?

**Review by sgugger — COMMENTED** [review]
Thanks for the PR! Got a few comments on my side.

File: src/transformers/trainer.py (hunk 9 above)
Comment (TevenLeScao) [reply src/transformers/trainer.py:1365]: Agree and will move them out, do you think we should keep a redirection to keep `distributed_concat` backwards-compatible?

File: src/transformers/trainer.py (hunk 11 above)
Comment (TevenLeScao) [reply src/transformers/trainer.py:1399]: yep, changed it, allows us to save a few lines in the main method too

File: src/transformers/trainer.py (hunk 8 above)
Comment (TevenLeScao) [reply src/transformers/trainer.py]: agreed, done

File: src/transformers/trainer.py [hunk 12]
```diff
@@ -468,7 +473,11 @@ def setup_wandb(self):
             logger.info(
                 'Automatic Weights & Biases logging enabled, to disable set os.environ["WANDB_DISABLED"] = "true"'
             )
-            combined_dict = {**self.model.config.to_dict(), **self.args.to_sanitized_dict()}
+            try:
+                combined_dict = {**self.model.config.to_dict(), **self.args.to_sanitized_dict()}
+            except AttributeError:
+                # in case the model has no config
+                combined_dict = {**self.args.to_sanitized_dict()}
```
Comment (LysandreJik) [inline src/transformers/trainer.py:480]: Is there an example of a model without a configuration?

File: src/transformers/trainer.py [hunk 13]
```diff
@@ -638,13 +647,16 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
...
+        self.total_flos = 0
         epochs_trained = 0
         steps_trained_in_current_epoch = 0
         # Check if continuing training from a checkpoint
         if model_path is not None:
             # set global_step to global_step of last saved checkpoint from model path
             try:
                 self.global_step = int(model_path.split("-")[-1].split(os.path.sep)[0])
+                self.total_flos = getattr(model.config, "total_flos", 0)
+
```
Comment (LysandreJik) [inline src/transformers/trainer.py:659]: wouldn't this fail if the model didn't have a config?

File: src/transformers/trainer.py (hunk 11 above)
Comment (LysandreJik) [reply src/transformers/trainer.py:1400]: We can remove the docstring as well

**Review by LysandreJik — APPROVED** [review]
LGTM, left a few comments.

File: src/transformers/trainer.py (hunk 12 above)
Comment (TevenLeScao) [reply src/transformers/trainer.py:480]: Ah yes, it's something @sgugger mentioned as well - when writing for `Trainer` we assume that the model is a `PretrainedModel`, but the one in the test doesn't inherit from `PretrainedModel` which is why I put this in. @julien-c also liked the idea of `Trainer` being domain-agnostic (eg not only NLP for example) so I figured might as well put this line in since it isn't expensive.

**Comment by LysandreJik** [comment]
I think even with domain-agnostic models we'd like to keep the configuration, no? I'm not sure the trainer would behave correctly without a configuration, so if we want to remove the dependency towards configurations, we might as well do it all at once, right?

Would the goal be to have the trainer accept all `nn.Module`s?

**Comment by sgugger** [comment]
Like agreed upon internally, we will move to Trainer accepting models instantiating a base abstractclass/conforming to some protocol. I think the config will be in the required field but have to work a bit more on this to be sure.

In any case, this is work for a subsequent PR :-)
